### PR TITLE
DI-160 Configure Api Gateway

### PR DESCRIPTION
**Pull request by rohith1122**
## Link to JIRA Ticket

- https://nhsd-jira.digital.nhs.uk/browse/DI-160

## Description

Configured API Gateway using serverless framework
### Noteworthy Changes
- Added new lambda "event-receiver" to receive the messages from NHS UK and this Lambda output the message in cloud watch. 
-  Reading ApiKey value from AWS Secret Manager  Service
-  Configured  HTTP event trigger to  lambda (event receiver) with an enpoint '/dos-integration/api/v1/nhsuk-event-receiver/{type}'  where type can be pharmacy/dental

## Type of change

Tick relevant ticket types

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please tick the testing that has been completed

- [ ] Unit tests

## Developer Checklist

- [x] I have performed a self-review of my own code
- [x] I have run the [code formatting checks](../README.md#code-quality)
- [x] I have run the [code quality checks](../README.md#code-quality)
- [ ] New code meets [standards](https://nhsd-confluence.digital.nhs.uk/display/DI/DI+Ways+of+Working) agreed by the team
- [ ] Unit test code coverage is at or above 80%
- [ ] New and existing unit tests pass locally with my changes
- [ ] Tests have added that prove my fix is effective or that my feature works
- [ ] I have made corresponding changes to the documentation

## Code Reviewer Checklist

- [x] I have run the unit tests and they run correctly
- [x] I can confirm the changes have been tested or approved by a tester

### CODE REVIEW

File: deployment/serverless.yml [hunk 1]
```diff
@@ -50,9 +58,36 @@ plugins:
 #           - 'uec-pu-apu-2429-rds'
 
 functions:
+  event-receiver:
+    image: ${env:AWS_SAME_ACCOUNT_DOCKER_REGISTRY}/event-receiver:latest
```
Comment (JackPlowman) [inline deployment/serverless.yml:62]: Can the version for both event-receiver and event-sender be set to ${env:VERSION}. This is so multiple people can deploy and we can deploy non latest versions
```suggestion
    image: ${env:AWS_SAME_ACCOUNT_DOCKER_REGISTRY}/event-receiver:${env:VERSION}
```

File: application/event_receiver/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+requests
```
Comment (JackPlowman) [inline application/event_receiver/requirements.txt:1]: Where is requests used? If it isn't can it be removed please

**Review by JackPlowman — CHANGES_REQUESTED** [review]
Sorry to come back with more comments. Please can a short section be added to the readme to declare a API Key must be generated prior to external API-Gateways being set up.

File: deployment/serverless.yml [hunk 3]
```diff
@@ -19,6 +19,14 @@ provider:
     Product: ${env:PROJECT_GROUP_SHORT}
     Programme: ${env:PROGRAMME}
     Service: ${env:PROJECT_GROUP_SHORT}
+  apiGateway:
+    apiKeySourceType: HEADER
+    apiKeys:
+      - name: NHS_UK_API_KEY
+        value: ${self:custom.secret.NHS_UK_API_KEY}
+custom:
+  secret: ${ssm:/aws/reference/secretsmanager/uec-dos-int-dev-api-key}
```
Comment (JackPlowman) [inline deployment/serverless.yml:28]: Can the secret key name and key be variables and stored in task.mk (build/automation/var/profile/task.mk)

**Review by JackPlowman — CHANGES_REQUESTED** [review]
Please note when this API-Gateway is set to private it means connection requires api key not that it is private to the VPC. The team need to have a discussion on this.
Suggestions for enhanced security/privacy
- https://medium.com/graymatrix/aws-private-apis-using-serverless-framework-785de73b2447
- (Add policy to API Gateway to block non vpc traffic) https://github.com/serverless/serverless/pull/5080

**Comment by JackPlowman** [comment]
I believe we need to make this a variable. Task is private, Dev up for debate, Demo/Live: Regional
`endpointType: regional # Optional endpoint configuration for API Gateway REST API. Default is Edge.`

File: deployment/serverless.yml (hunk 1 above)
Comment (rohith1122) [reply deployment/serverless.yml]: updated

File: application/event_receiver/requirements.txt (hunk 2 above)
Comment (rohith1122) [reply application/event_receiver/requirements.txt]: done

File: deployment/serverless.yml [hunk 4]
```diff
@@ -66,7 +77,17 @@ functions:
...
+    events:
+      - http:
+        path: /dos-integration/api/v1/nhsuk-event-receiver/{type}
+        request:
+          parameters:
+            paths:
+              type: true # pharmacy /dental
+        method: post
+        private: true
+    # tracing:
```
[the remark this answers is not in the document]
Comment (JackPlowman) [reply deployment/serverless.yml]: No problem, I looked at the documentation and is v1 api gateway

File: deployment/serverless.yml (hunk 3 above)
Comment (rohith1122) [reply deployment/serverless.yml]: done

**Review by JackPlowman — DISMISSED** [review]
This needs to be rebased and then the secrets issue should go away

File: Makefile [hunk 5]
```diff
@@ -147,34 +118,82 @@ common-code-remove: ### Remove common code from lambda direcory - mandatory: LAM
...
+	-e POWERTOOLS_METRICS_NAMESPACE="dos-integration" \
+	-e POWERTOOLS_SERVICE_NAME="event-sender" \
+	" \
+	CONTAINER="event-sender"
+
+event-sender-trigger: ### Trigger event sender lambda
+	curl -XPOST "http://localhost:9000/2015-03-31/functions/function/invocations" -d '{"message": "hello world", "username": "lessa"}'
+
 event-sender-run: ### A rebuild and restart of the event sender lambda.
-	make stop
```
Comment (JackPlowman) [inline Makefile:167]: Can this make target be reverted please

File: build/automation/var/profile/dev.mk [hunk 6]
```diff
@@ -12,10 +12,5 @@ DOS_API_GATEWAY_PASSWORD := $(or $(DOS_API_GATEWAY_PASSWORD), "")
 DOS_API_GATEWAY_REQUEST_TIMEOUT := 30
 DOS_API_GATEWAY_URL := $(or $(DOS_API_GATEWAY_URL), "//")
 
-TF_VAR_api_gateway_api_key_name := $(PROJECT_ID)-$(PROFILE)-api-key
```
Comment (JackPlowman) [inline build/automation/var/profile/dev.mk:15]: Can all these variables be re-added please

File: build/automation/var/profile/task.mk [hunk 7]
```diff
@@ -12,7 +12,5 @@ DOS_API_GATEWAY_URL := $(or $(DOS_API_GATEWAY_URL), "//")
 # ==============================================================================
 # Infrastructure variables (Terraform, Serverless, etc)
 
-TF_VAR_api_gateway_api_key_name := $(PROJECT_ID)-$(ENVIRONMENT)-api-key
-TF_VAR_nhs_uk_api_key_key := NHS_UK_API_KEY
```
Comment (JackPlowman) [inline build/automation/var/profile/task.mk:16]: Can all these variables be re-added please

File: deployment/serverless.yml [hunk 8]
```diff
@@ -19,48 +19,91 @@ provider:
...
+        #   Action: 'TODO' #This need to be restricted when used
+        # - Effect: 'Deny'
+        #   Resource: 'TODO'
+        #   Action: 'TODO'
 
 plugins:
   - serverless-vpc-discovery
   - serverless-localstack
 
-custom:
```
Comment (JackPlowman) [inline deployment/serverless.yml:27]: Can this section be reset

File: deployment/serverless.yml [hunk 9]
```diff
@@ -19,48 +19,91 @@ provider:
...
-  vpcDiscovery:
```
Comment (JackPlowman) [inline deployment/serverless.yml:28]: Also the other custom section can be merged into this custom

File: deployment/serverless.yml [hunk 10]
```diff
@@ -19,48 +19,91 @@ provider:
...
+                type: true # pharmacy /dental
+          method: post
+          private: true
   event-sender:
     image: ${env:AWS_SAME_ACCOUNT_DOCKER_REGISTRY}/event-sender:${env:VERSION}
     architecture: x86_64
     name: ${env:PROJECT_ID}-${sls:stage}-event-sender
     description: ${sls:stage} Event Sender lambda
     memorySize: 512
-    timeout: 60
```
Comment (JackPlowman) [inline deployment/serverless.yml:49]: Can everything except tracing be re-added to this file

**Review by JackPlowman — CHANGES_REQUESTED** [review]
Sorry Rohit, but the rebase seems to have screwed up your branch

File: Makefile [hunk 11]
```diff
@@ -4,83 +4,54 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
+	make python-virtualenv
+	pip install -r application/requirements.txt
+
+build-dev: # Build dev requirements
 	make docker-build NAME=serverless
 	make serverless-requirements
 	make python-requirements
 
 build: # Build lambdas
-	make event-sender-build AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile]: Resolved

File: Makefile [hunk 12]
```diff
@@ -4,83 +4,54 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
-
-start: # Stop project
-	make project-start AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile]: Resolved

File: Makefile [hunk 13]
```diff
@@ -4,83 +4,54 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
+	make event-sender-build
+	make event-receiver-build
 
-stop: # Stop project
-	make project-stop AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile]: Resolved

File: Makefile [hunk 14]
```diff
@@ -4,83 +4,54 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
+stop: project-stop # Stop project
 
 restart: stop start # Restart project
 
 log: project-log # Show project logs
 
 deploy: # Deploys whole project - mandatory: PROFILE
-	eval "$$(make populate-deployment-variables)"
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile:26]: Resolved

File: Makefile [hunk 15]
```diff
@@ -4,83 +4,54 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
-	make terraform-apply-auto-approve STACKS=lambda-security-group
-	make serverless-deploy
-
-sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit hash-timestamp]
-	eval "$$(make populate-deployment-variables)"
```
[the remark this answers is not in the document]
Comment (rohith1122) [reply Makefile:31]: Resolved

File: build/automation/var/profile/dev.mk (hunk 6 above)
Comment (rohith1122) [reply build/automation/var/profile/dev.mk]: Resolved

File: build/automation/var/profile/task.mk (hunk 7 above)
Comment (rohith1122) [reply build/automation/var/profile/task.mk:16]: Resolved

File: deployment/serverless.yml (hunk 8 above)
Comment (rohith1122) [reply deployment/serverless.yml:27]: Resolved

File: deployment/serverless.yml (hunk 9 above)
Comment (rohith1122) [reply deployment/serverless.yml:28]: Resolved

File: deployment/serverless.yml (hunk 10 above)
Comment (rohith1122) [reply deployment/serverless.yml:49]: Resolved

**Comment by JackPlowman** [comment]
I'm going to resolve all comments then rereview

File: Makefile [hunk 16]
```diff
@@ -33,54 +37,29 @@ sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit has
 
 undeploy: # Undeploys whole project - mandatory: PROFILE
 	make serverless-remove VERSION="any"
```
Comment (JackPlowman) [inline Makefile:39]: Add this line here please
```suggestion
	make serverless-remove VERSION="any"
	make terraform-destroy-auto-approve STACKS=lambda-security-group
```

File: Makefile [hunk 17]
```diff
@@ -33,54 +37,29 @@ sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit has
...
-		echo "export DOS_API_GATEWAY_USERNAME=$$(make -s secret-get-existing-value NAME=$(DOS_DEPLOYMENT_SECRETS) KEY=$(DOS_API_GATEWAY_USERNAME_KEY))"
-		echo "export DOS_API_GATEWAY_PASSWORD=$$(make -s secret-get-existing-value NAME=$(DOS_DEPLOYMENT_SECRETS) KEY=$(DOS_API_GATEWAY_PASSWORD_KEY))"
-	fi
-
 python-requirements: # Installs whole project python requirements
 	make docker-run-tools \
 		CMD="pip install -r requirements.txt -r requirements-dev.txt -r event_sender/requirements.txt" \
 		DIR=./application
 
 unit-test: # Runs whole project unit tests
```
Comment (JackPlowman) [inline Makefile:51]: Please revert this make target
```suggestion
unit-test: # Runs whole project unit tests
	make -s docker-run-tools \
	CMD="python -m pytest --cov=." \
This conversation was marked as resolved by JackPlowman
 Show conversation
	DIR=./application \
	ARGS=" \
		-e POWERTOOLS_LOG_DEDUPLICATION_DISABLED="1" \
		--volume $(APPLICATION_DIR)/event_sender:/tmp/.packages/event_sender \
		"```

File: Makefile [hunk 18]
```diff
@@ -33,54 +37,29 @@ sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit has
...
-
-component-test: # Runs whole project component tests
-	make -s docker-run-tools \
-	CMD="python -m behave" \
-	DIR=test/component \
-	ARGS=" \
-		-e MOCKSERVER_URL=$(MOCKSERVER_URL) \
-		-e EVENT_SENDER_FUNCTION_URL=$(EVENT_SENDER_FUNCTION_URL)\
-		"
+	make python-code-coverage DIR=./application \
```
Comment (JackPlowman) [inline Makefile:58]: Please revert this make target

File: Makefile [hunk 19]
```diff
@@ -33,54 +37,29 @@ sls-only-deploy: # Deploys all lambdas - mandatory: PROFILE, VERSION=[commit has
...
+	DIR=application \
+	ARGS="-e POWERTOOLS_TRACE_DISABLED=1"
 
 coverage-report: # Runs whole project coverage unit tests
-	make python-code-coverage DIR=$(APPLICATION_DIR_REL) \
-	ARGS=" \
-		--volume $(APPLICATION_DIR)/event_sender:/tmp/.packages/event_sender \
-		"
-
-component-test: # Runs whole project component tests
```
Comment (JackPlowman) [inline Makefile:70]: Please re-add this make target

**Review by JackPlowman — CHANGES_REQUESTED** [review]
A few things still to be fixed. Also I have unresolved issues that still need to be dealt with

File: Makefile [hunk 20]
```diff
@@ -5,50 +5,72 @@ include $(abspath $(PROJECT_DIR)/build/automation/init.mk)
...
 
 build-dev: # Build dev requirements
 	make docker-build NAME=serverless
 	make serverless-requirements
 	make python-requirements
 
 build: # Build lambdas
-	make event-sender-build
-	make event-receiver-build
+	make event-sender-build AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
```
Comment (JackPlowman) [inline Makefile]: I missed this one in my review but can you add this
```suggestion
  make event-sender-build AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
  make event-receiver-build AWS_ACCOUNT_ID_MGMT=$(AWS_ACCOUNT_ID_NONPROD)
```

**Review by JackPlowman — DISMISSED** [review]
LGMT

**Review by nat-darke-nhs — DISMISSED** [review]

**Review by nat-darke-nhs — DISMISSED** [review]

**Review by JackPlowman — APPROVED** [review]
LGTM 👍

**Review by nat-darke-nhs — APPROVED** [review]

**Review by JackPlowman — APPROVED** [review]

**Pull request merged**